File: bundle/edu.gemini.shared.gui/src/main/scala/edu/gemini/shared/Platform.scala
package edu.gemini.shared

import java.awt.Toolkit
import java.awt.event.InputEvent
import java.net.URL
import java.util.logging.Logger
import javax.swing.{InputMap, JTextArea, JTextField, KeyStroke}

object Platform extends Enumeration {

  val Mac, Windows, Other = Value

  private val LOGGER = Logger.getLogger(getClass.getName)

  lazy val current = {
    def is(s: String) = System.getProperty("os.name").contains(s)
    if (is("Mac")) Mac else if (is("Windows")) Windows else Other
  }

  def IS_MAC = current == Mac
  def IS_WINDOWS = current == Windows

  lazy val MENU_ACTION_MASK = Toolkit.getDefaultToolkit.getMenuShortcutKeyMask
  lazy val TOGGLE_ACTION_MASK = if (current == Mac) InputEvent.META_DOWN_MASK else InputEvent.CTRL_DOWN_MASK

  if (current == Mac) {

    val CTRL = InputEvent.CTRL_MASK | InputEvent.CTRL_DOWN_MASK

    def fix(omap: Option[InputMap]) {
      for {

      // Base case exits on None
        map <- omap

        // Do the recursive call here
        _ = fix(Option(map.getParent))

        // Get the keystrokes
        keys <- Option(map.keys)
        ks <- keys

        // And new modifiers
        mod = ks.getModifiers if (mod & CTRL) != 0
        newMod = (mod & ~CTRL) | MENU_ACTION_MASK
        newKs = KeyStroke.getKeyStroke(ks.getKeyCode, newMod, ks.isOnKeyRelease)
        action = map.get(ks)

      } {

        // Update the map
        map.remove(ks)
        map.put(newKs, action)

      }
    }

    LOGGER.info("Repairing edit component keymap for Mac OS X.")
    Seq(new JTextArea, new JTextField) foreach { o => fix(Option(o.getInputMap)) }

  }

  /*
   * Derived from http://www.javaworld.com/javaworld/javatips/jw-javatip66.html
   */
  def displayURL(url: URL) { displayURL(url.toString) }
  def displayURL(url: String) {
    current match {
      case Windows => exec("rundll32 url.dll,FileProtocolHandler " + url)
      case Mac     => exec("/usr/bin/open %s" format url)
      case _       => if (exec("firefox -remote openURL(%s)" format url).waitFor != 0) exec("firefox " + url)
    }
  }

  private def exec(s: String) = sys.runtime.exec(s)

}

File: bundle/edu.gemini.qpt.client/src/main/java/edu/gemini/qpt/ui/util/Platform.java
package edu.gemini.qpt.ui.util;

import java.awt.Toolkit;
import java.awt.event.InputEvent;
import java.io.IOException;
import java.net.URL;
import java.util.logging.Level;
import java.util.logging.Logger;

import javax.swing.InputMap;
import javax.swing.JTextArea;
import javax.swing.JTextField;
import javax.swing.KeyStroke;
import javax.swing.text.JTextComponent;

public class Platform {

    private static final Logger LOGGER = Logger.getLogger(Platform.class.getName());
    
    public static final boolean IS_MAC = System.getProperty("os.name").contains("Mac");
    public static final boolean IS_WINDOWS = System.getProperty("os.name").contains("Windows");
    
    public static final int MENU_ACTION_MASK = Toolkit.getDefaultToolkit().getMenuShortcutKeyMask(); 
    
    public static final int TOGGLE_ACTION_MASK = 
        IS_MAC ? InputEvent.META_DOWN_MASK : InputEvent.CTRL_DOWN_MASK;
    
    static {

        // Make sure all the editor keys are mapped to ACTION_MASK instead of just using
        // CTRL. This only changes things on the Mac.
        
        if (IS_MAC) {
            
            LOGGER.fine("Repairing edit component keymap for Mac OS X.");
            JTextComponent[] jtcs = new JTextComponent[] {
                new JTextArea(),
                new JTextField(),
            };
            
            int CTRL = InputEvent.CTRL_MASK | InputEvent.CTRL_DOWN_MASK;
            for (JTextComponent jtc: jtcs) {
                for (InputMap map = jtc.getInputMap(); map != null; map = map.getParent()) {
                    if (map.keys() != null) {
                        for (KeyStroke ks: map.keys()) {
                            int mod = ks.getModifiers();
                            if ((mod & CTRL) != 0) {
                                int newMod = (mod & ~CTRL) | MENU_ACTION_MASK;
                                KeyStroke newKs = KeyStroke.getKeyStroke(ks.getKeyCode(), newMod, ks.isOnKeyRelease());
                                Object action = map.get(ks);
                                map.remove(ks);
                                map.put(newKs, action);
                            }
                        }
                    }
                }
            }
            
        }            
        
    }
    
    /*
     * Derived from http://www.javaworld.com/javaworld/javatips/jw-javatip66.html
     */
    public static boolean displayURL(String url) throws IOException {
        try {
            if (IS_WINDOWS) {
                Runtime.getRuntime().exec("rundll32 url.dll,FileProtocolHandler " + url);
            } else if (IS_MAC) {
                Runtime.getRuntime().exec("/usr/bin/open " + url);                
            } else {
                Process p = Runtime.getRuntime().exec("firefox -remote openURL(" + url + ")");
                try {
                    int exitCode = p.waitFor();
                    if (exitCode != 0)
                        Runtime.getRuntime().exec("firefox " + url);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    throw new IOException("Operation was interrupted.");
                }                
            }
            return false;
        } catch (IOException ioe) {
            LOGGER.log(Level.WARNING, "Trouble opening " + url, ioe);
            throw ioe;
        }
    }
    
    public static boolean displayURL(URL url) throws IOException {
        return displayURL(url.toString());
    }

    
//    public static void main(String[] args) {
//        
//        int mod = InputEvent.CTRL_DOWN_MASK | InputEvent.SHIFT_DOWN_MASK;
//        
//        System.out.println(Integer.toString(mod, 2));
//        
//        mod = (mod & ~InputEvent.CTRL_DOWN_MASK) | ACTION_MASK;
//        
//        System.out.println(Integer.toString(mod, 2));
//        
//    }

}
